### PR TITLE
Implemented #62

### LINKED ISSUE #62: Line numbers
As a user, who wants to talk about the content I read, I need to quickly determine where I am in the text in order to communicate about the given text.

**Acceptance criteria:**
- [x] There is an indication on screen in which line the cursor currently is
- [ ] optional: Next to each line the line number is shown
- [x] optional: There is an indication on screen what the position in the line is (number of character)

**Estimated Effort:** 5

### PR DESCRIPTION
**Pull request by j-hellenberg**
Also fixed some tests, footer is now updated on every action instead on every save.
- Discussion with customer (perfomance impact)?

### CODE REVIEW

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st [hunk 1]
```diff
@@ -0,0 +1,12 @@
+document information
+positionInTextAtIndex: anIndex
+
+	| lineCounter currentLine positionInLine |
+
+	lineCounter := 0.
+	currentLine := self displayedLines 
+		detect: [: each | lineCounter := lineCounter + 1. each first <= (anIndex + 1) and: [each last >= (anIndex + 1)]]
```
Comment (frcroth) [inline packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st]: line too long

File: packages/RichTextEditing-Tests.package/RichTextEditorTests.class/instance/testUS04loadedTextIsInsertedIntoEditor.st [hunk 2]
```diff
@@ -8,4 +8,4 @@ testUS04loadedTextIsInsertedIntoEditor
 	self editorModel setText: self anotherText.
 	self loadDocumentFromSaver.
 	
-	self assert: self text equals: self editorModel text
\ No newline at end of file
+	self assert: self editorModel text equals: self text
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextEditorTests.class/instance/testUS04loadedTextIsInsertedIntoEditor.st]: trailing space

File: packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/multipleLineText.st [hunk 3]
```diff
@@ -0,0 +1,11 @@
+constants
+multipleLineText
+
+	^ 'This is an text
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/multipleLineText.st]: Typo, should be This is **a** text

File: packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st [hunk 4]
```diff
@@ -0,0 +1,7 @@
+unit-tests
+testCorrectLineColNumberAtArbitraryPositionInText
+
+	self enterStringOntoTextField: self multipleLineText.
+	self moveCaretToBeginningOfText.
+	self moveCaretMultipleTimesRight: (self standardText size + 23).
+	self assert: (self editorModel currentPositionInText) = (7@3)
```
Comment (frcroth) [inline packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st]: use cascading

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st (hunk 1 above)
Comment (j-hellenberg) [reply packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st]: What would correct formatting look like here?

File: packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st (hunk 4 above)
Comment (j-hellenberg) [reply packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st]: True

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st (hunk 1 above)
Comment (frcroth) [reply packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st]: perhaps new line before `and:`

**Comment by oleschl** [comment]
Performance at least on my laptop seems fine most of the time. I tested a text with about 10000 words and performance was okay when adding new text in between old text. However, when adding new text to the end it takes more than a second for a single character. Same happens in the standard workspace.

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st [hunk 5]
```diff
@@ -0,0 +1,14 @@
+document information
+positionInTextAtIndex: anIndex
+
+	| lineCounter currentLine positionInLine |
+
+	lineCounter := 0.
+	currentLine := self displayedLines 
+		detect:
```
Comment (oleschl) [inline packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:8]: Isn't this the collection count function?

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+	lineCounter := 0.
+	currentLine := self displayedLines 
+		detect: 
+			[: each | lineCounter := lineCounter + 1. 
+			each first <= (anIndex + 1) and: [each last >= (anIndex + 1)]] 
+		ifNone: [self displayedLines last].
+	positionInLine := anIndex - currentLine first + 2.
+	
+	^ positionInLine @ lineCounter
```
Comment (oleschl) [inline packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:14]: I'm not sure if I like the point here or if it's better to split it into two different methods

**Comment by oleschl** [comment]
> Also fixed some tests, footer is now updated on every action instead on every save.
> 
>     * Discussion with customer (perfomance impact)?

Kinda conflicts with the last time saved field (US86) which should be updated when saving

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st (hunk 5 above)
Comment (j-hellenberg) [reply packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:8]: It actually returns the first element for which the block evaluates to true

**Comment by j-hellenberg** [comment]
> > Also fixed some tests, footer is now updated on every action instead on every save.
> > ```
> > * Discussion with customer (perfomance impact)?
> > ```
> 
> Kinda conflicts with the last time saved field (US86) which should be updated when saving

But we can do an footer update on saving as well, can't we? I actually have not removed it, so I don't quite see the problem

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st (hunk 6 above)
Comment (j-hellenberg) [reply packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:14]: Good point, I also thought about that. Not quite sure what I prefer though

**Comment by j-hellenberg** [comment]
> Performance at least on my laptop seems fine most of the time. I tested a text with about 10000 words and performance was okay when adding new text in between old text. However, when adding new text to the end it takes more than a second for a single character. Same happens in the standard workspace.

Writing text fast produces some lag before it catches up on my local. The normal editor also has got that problem, but not quite as much I feel. Besides that, it is working fine

File: packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st (hunk 6 above)
Comment (sirkrypt0) [reply packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:14]: A point is fine imo, it represents the context better than two methods, as the position in the text is here really a position in form of x@y

File: packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st [hunk 7]
```diff
@@ -0,0 +1,8 @@
+unit-tests
+testCorrectLineColNumberAtArbitraryPositionInText
+
+	self 
+		enterStringOntoTextField: self multipleLineText;
```
Comment (sirkrypt0) [inline packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st:5]: You enter the `multipleLineText` here but move the caret right based on the `standardText size`. Is that intended and why?

**Review by sirkrypt0 — CHANGES_REQUESTED** [review]

**Review by frcroth — APPROVED** [review]

File: packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st (hunk 7 above)
Comment (j-hellenberg) [reply packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st:5]: Yea, kinda. The point is, that standardText is already entered in setup but the carret is still considerered to be on position 1. The multipleLineText is appended afterwards. Therefore, I have to move it by the size of stardardText first to get an reasonable test result.

File: packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+		moveCaretToBeginningOfText;
+		moveCaretMultipleTimesRight: (self standardText size + 22);
```
Comment (sirkrypt0) [inline packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st:7]: Where does that magic `22` come from?

**Review by sirkrypt0 — CHANGES_REQUESTED** [review]

**Review by sirkrypt0 — APPROVED** [review]

**Pull request merged**